lib/srv/handlers.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/sshutils"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// AuthHandlers are common authorization and authentication related handlers
+// used by the regular and forwarding server.
+type AuthHandlers struct {
+	// Server is the services.Server in the backend.
+	Server services.Server
+
+	// ProxyMode is if the server is running in proxy mode. Only applies to the
+	// regular server.
+	ProxyMode bool
+
+	// AuditLog is the service used to access Audit Log.
+	AuditLog events.IAuditLog
+
+	// AccessPoint is used to access the Auth Server.
+	AccessPoint auth.AccessPoint
+}
+
+// CheckAgentForward checks the certificate was signed by a valid certificate
+// authority, fetches the appropriate roles, and checks if agent forwarding
+// is allowed.
+func (h *AuthHandlers) CheckAgentForward(ctx *ServerContext) error {
+	cert, err := ctx.GetCertificate()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ca, err := h.authorityForCert(cert.SignatureKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	roles, err := h.fetchRoleSet(cert, ca, ctx.TeleportUser, ctx.ClusterName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := roles.CheckAgentForward(ctx.Login); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// KeyAuth implements SSH client authentication using public keys and is called
+// by the server every time the client connects
+func (h *AuthHandlers) KeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+	cid := fmt.Sprintf("conn(%v->%v, user=%v)", conn.RemoteAddr(), conn.LocalAddr(), conn.User())
+	fingerprint := fmt.Sprintf("%v %v", key.Type(), sshutils.Fingerprint(key))
+	log.Debugf("[SSH] %v auth attempt with key %v", cid, fingerprint)
```

Comment:
can you set `trace.Component` here instead of `[SSH]` as you are refactoring this chunk anyways.